lib/rspec-puppet/support.rb

```diff
@@ -279,6 +279,34 @@ def trusted_facts_hash(node_name)
       extensions
     end
 
+    def server_facts_hash
+      server_facts = {}
+
+      # Add our server version to the fact list
+      server_facts["serverversion"] = Puppet.version.to_s
+
+      # And then add the server name and IP
+      {"servername" => "fqdn",
+        "serverip" => "ipaddress"
+      }.each do |var, fact|
+        if value = Facter.value(fact)
+          server_facts[var] = value
+        else
+          warn "Could not retrieve fact #{fact}"
+        end
+      end
+
+      if server_facts["servername"].nil?
+        host = Facter.value(:hostname)
+        if domain = Facter.value(:domain)
+          server_facts["servername"] = [host, domain].join(".")
+        else
+          server_facts["servername"] = host
+        end
+      end
```

Comment:
How does the `$server_facts['environment']` key get assigned?  I don't see it here, but the (Puppet 5) spec tests pass so it looks like it happens somewhere.

Reply:
`$server_facts['environment']` gets set here: https://github.com/puppetlabs/puppet/blob/2b7b4ed6b692bf595d4f95012a6a132024c991cb/lib/puppet/node.rb#L163-L164, which in-turn is defined by rspec-puppet here: https://github.com/rodjek/rspec-puppet/blob/fb24311494b35978218ed2d77a5a1468a84f745a/lib/rspec-puppet/support.rb#L212

Reply:
The environment server fact is set when the server facts are added to the node. The value is forced into whatever environment the compiler is using, so `let(:environment)` will set the environment server fact as well.